File: app/src/main/java/com/lumiyaviewer/lumiya/render/shaders/ShaderPreprocessor.java
package com.lumiyaviewer.lumiya.render.shaders;

import com.google.common.base.Objects;
import com.google.common.collect.ImmutableMap;
import java.io.BufferedReader;
import java.io.IOException;
import java.util.Map;
import java.util.Map.Entry;
import javax.annotation.Nullable;

public class ShaderPreprocessor {
    private final ImmutableMap<String, String> definedMacros;

    public ShaderPreprocessor(Map<String, String> map) {
        this.definedMacros = ImmutableMap.copyOf(map);
    }

    @Nullable
    private String processCode(BufferedReader bufferedReader, @Nullable StringBuilder stringBuilder) throws IOException {
        String str = null;
        while (true) {
            String readLine = bufferedReader.readLine();
            if (readLine == null) {
                return str;
            }
            readLine = readLine.trim();
            if (readLine.startsWith("#endif") || readLine.startsWith("#else")) {
                return readLine;
            }
            if (readLine.startsWith("#ifdef") || readLine.startsWith("#ifndef")) {
                boolean startsWith = readLine.startsWith("#ifdef");
                boolean containsKey = this.definedMacros.containsKey(readLine.substring(readLine.indexOf(32)).trim());
                String processCode = processCode(bufferedReader, startsWith == containsKey ? stringBuilder : null);
                if (Objects.equal(processCode, "#else")) {
                    processCode = processCode(bufferedReader, startsWith != containsKey ? stringBuilder : null);
                }
                if (!Objects.equal(processCode, "#endif")) {
                    throw new IOException("#endif expected");
                }
            } else if (stringBuilder != null) {
                String str2 = readLine;
                for (Entry entry : this.definedMacros.entrySet()) {
                    str2 = str2.replace((CharSequence) entry.getKey(), (CharSequence) entry.getValue());
                }
                stringBuilder.append(str2).append("\r\n");
                str = str2;
            }
            str = readLine;
        }
    }

    public String processCode(BufferedReader bufferedReader) throws IOException {
        StringBuilder stringBuilder = new StringBuilder();
        processCode(bufferedReader, stringBuilder);
        return stringBuilder.toString();
    }
}

File: app/src/main/java/com/lumiyaviewer/lumiya/render/shaders/ShaderPreprocessor.kt
package com.lumiyaviewer.lumiya.render.shaders

import com.google.common.base.Objects
import com.google.common.collect.ImmutableMap
import java.io.BufferedReader
import java.io.IOException
import javax.annotation.Nullable

/**
 * Preprocessor for shader source code.
 * Handles #ifdef, #ifndef, #else, #endif directives and macro substitution.
 */
class ShaderPreprocessor(map: Map<String, String>) {
    private val definedMacros: ImmutableMap<String, String> = ImmutableMap.copyOf(map)

    /**
     * Process shader code with preprocessor directives.
     * @param bufferedReader Source code reader
     * @return Processed code as string
     * @throws IOException if reading fails or syntax errors occur
     */
    @Throws(IOException::class)
    fun processCode(bufferedReader: BufferedReader): String {
        val stringBuilder = StringBuilder()
        processCode(bufferedReader, stringBuilder)
        return stringBuilder.toString()
    }

    @Nullable
    @Throws(IOException::class)
    private fun processCode(
        bufferedReader: BufferedReader,
        @Nullable stringBuilder: StringBuilder?
    ): String? {
        var lastLine: String? = null
        
        while (true) {
            val readLine = bufferedReader.readLine() ?: return lastLine
            val trimmedLine = readLine.trim()
            
            // Check for directive end markers
            if (trimmedLine.startsWith("#endif") || trimmedLine.startsWith("#else")) {
                return trimmedLine
            }
            
            // Handle conditional compilation
            if (trimmedLine.startsWith("#ifdef") || trimmedLine.startsWith("#ifndef")) {
                val isIfdef = trimmedLine.startsWith("#ifdef")
                val macroName = trimmedLine.substring(trimmedLine.indexOf(' ')).trim()
                val isDefined = definedMacros.containsKey(macroName)
                
                // Process the conditional block
                var nextDirective = processCode(
                    bufferedReader,
                    if (isIfdef == isDefined) stringBuilder else null
                )
                
                // Handle #else block
                if (Objects.equal(nextDirective, "#else")) {
                    nextDirective = processCode(
                        bufferedReader,
                        if (isIfdef != isDefined) stringBuilder else null
                    )
                }
                
                // Verify proper closing
                if (!Objects.equal(nextDirective, "#endif")) {
                    throw IOException("#endif expected")
                }
            } else if (stringBuilder != null) {
                // Process active code - perform macro substitution
                // Replace only whole tokens (macro names) using word boundaries
                var processedLine = trimmedLine
                for (entry in definedMacros.entries) {
                    val pattern = Regex("\\b" + Regex.escape(entry.key) + "\\b")
                    processedLine = pattern.replace(processedLine, entry.value)
                }
                stringBuilder.append(processedLine).append("\r\n")
                lastLine = processedLine
            }
            
            lastLine = trimmedLine
        }
    }
}
